### PR TITLE
Make Multiplexer inherit filter chains analysis mode

### LINKED ISSUE #50554: Reload search analyzers does not work correctly with multiplexer filter (showstopper)
**Elasticsearch version**: Version: 7.5.1

**Plugins installed**: []

**JVM version**: 13.0.1

**OS version**: Windows 10 Pro

**Description of the problem including expected versus actual behavior**: When the synonym_graph filter is not used directly in an analyzer but is included in multiplexer filter (for example to allow combination with word_delimiter_graph), _reload_search_analyzers does not correctly find and reload appropriate analyzer.

**Steps to reproduce**:

 1. Create index
```
PUT /my_index
{
  "settings": {
    "index": {
      "analysis": {
        "analyzer": {
          "my_synonyms_only": {
            "tokenizer": "whitespace",
            "filter": [
              "synonym"
            ]
          },
          "my_synonyms": {
            "tokenizer": "whitespace",
            "filter": [
              "multiplexer"
            ]
          }
        },
        "filter": {
          "synonym": {
            "type": "synonym_graph",
            "synonyms_path": "analysis/synonyms.txt",
            "updateable": true
          },
          "delimiter": {
            "type": "word_delimiter_graph",
            "preserve_original": true
          },
          "multiplexer": {
            "type": "multiplexer",
            "filters": [
              "delimiter",
              "synonym"
            ]
          }
        }
      }
    }
  }
}
```
 2. Try to reload search analyzers
```
POST /my_index/_reload_search_analyzers
```

 3. Only the analyzer which uses synonym filter directly is updated. Analyzer with multiplexer is not.
```
  "reload_details" : [
    {
      "index" : "my_index",
      "reloaded_analyzers" : [
        "my_synonyms_only"
      ],
      "reloaded_node_ids" : [
        "ex70i64yRtar5gMPncOUTw"
      ]
    }
  ]
```
**4.  I would expect the my_synonyms analyzer to be reloaded also**

### PR DESCRIPTION
**Pull request by cbuescher**
Currently, if an updateable synonym filter is included in a multiplexer filter, it is not reloaded 
via the `_reload_search_analyzers` because the multiplexer itself doesn't pass on the analysis mode 
of the filters it contains, so its not recognized as "updateable" in itself. Instead we can check and merge
the `AnalysisMode`  settings of all filters in the multiplexer and use the resulting mode (e.g. search-time only)
for the multiplexer itself, thus making any synonym filters contained in it reloadable. This, of course, will also
make the analyzers using the multiplexer be usable at search-time only.

Marking this as WIP for now to get some test runs and for discussion.

Closes #50554

### CODE REVIEW

**Comment by cbuescher** [comment]
@elasticmachine  run elasticsearch-ci/bwc
@elasticmachine  run elasticsearch-ci/default-distro

**Comment by cbuescher** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 package org.elasticsearch.index.mapper;
 
 import com.carrotsearch.hppc.ObjectHashSet;
+
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java]: nit: I think this is a leftover?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java [hunk 2]
```diff
@@ -126,6 +122,74 @@ public void testSynonymsUpdateable() throws FileNotFoundException, IOException {
...
+            Files.createDirectory(configDir);
+        }
+        Path synonymsFile = configDir.resolve(synonymsFileName);
+        if (Files.exists(synonymsFile) == false) {
+            Files.createFile(synonymsFile);
+        }
+        try (PrintWriter out = new PrintWriter(
+                new OutputStreamWriter(Files.newOutputStream(synonymsFile, StandardOpenOption.WRITE), StandardCharsets.UTF_8))) {
+            out.println("foo, baz");
+        }
```
Comment (romseygeek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java]: Given that this is reused in three different tests now, I wonder if it's worth extracting it as a separate method?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java [hunk 3]
```diff
@@ -143,20 +207,15 @@ public void testUpdateableSynonymsRejectedAtIndexTime() throws FileNotFoundExcep
...
-                .put("analysis.filter.synonym_filter.type", "synonym")
-                .put("analysis.filter.synonym_filter.updateable", "true")
-                .put("analysis.filter.synonym_filter.synonyms_path", synonymsFileName))
-                .addMapping("_doc", "field", "type=text,analyzer=" + analyzerName).get());
-
-        assertEquals(
-                "Failed to parse mapping: analyzer [my_synonym_analyzer] "
-                + "contains filters [synonym_filter] that are not allowed to run in all mode.",
-                ex.getMessage());
+        MapperException ex = expectThrows(MapperException.class,
```
Comment (romseygeek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java]: This compression of multiple settings onto single lines seems to me to make test more difficult to follow, can we keep the indentation as it was before?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java [hunk 4]
```diff
@@ -126,6 +122,74 @@ public void testSynonymsUpdateable() throws FileNotFoundException, IOException {
...
+        assertTrue(tokens.contains("foo"));
+        assertTrue(tokens.contains("baz"));
+        assertTrue(tokens.contains("buzz"));
+
+        response = client().prepareSearch(indexName).setQuery(QueryBuilders.matchQuery("field", "baz")).get();
+        assertHitCount(response, 1L);
+        response = client().prepareSearch(indexName).setQuery(QueryBuilders.matchQuery("field", "buzz")).get();
+        assertHitCount(response, 1L);
+    }
+
```
Comment (romseygeek) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java]: Maybe also add a test asserting that a multiplexer containing updateable synonyms is rejected as an index-time analyzer as well?

**Review by romseygeek — CHANGES_REQUESTED** [review]
Thanks @cbuescher, looks good; there's one more test that I think is worth adding.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java (hunk 3 above)
Comment (cbuescher) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/ReloadSynonymAnalyzerTests.java]: Good point, this was not intended, must be the new Eclipse installation doing some autoformatting.

**Comment by cbuescher** [comment]
@romseygeek thanks for the review, your comments should be adressed now

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @cbuescher!

**Pull request merged**